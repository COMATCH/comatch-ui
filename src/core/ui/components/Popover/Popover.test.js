import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import '../../../config/tests/setup';

import { Popover } from './index';


describe('Popover', () => {
    it('should render correctly', () => {
        const popover = renderer.create(<Popover />).toJSON();
        expect(popover).toMatchSnapshot();
    });

    it('should have the correct default props', () => {
        const popover = mount(<Popover />);
        expect(popover.prop('children')).toEqual([]);
        expect(popover.prop('toggle')).toEqual(null);
        expect(popover.prop('open')).toEqual(null);
    });

    describe('passed props', () => {
        describe('custom toggel', () => {
            it('should render correctly', () => {
                const customToggle = <button icon='customIcon'>Hover me</button>;
                const popover = mount(<Popover toggle={customToggle} />);
                const button = popover.find('button');

                expect(button.prop('icon')).toEqual('customIcon');
                expect(button.text()).toEqual('Hover me');
            });
        });

        describe('position', () => {
            it('should render properly', () => {
                const popover = mount(<Popover position='left' />);
                expect(popover.prop('position')).toEqual('left');
            });
        });
    });

    describe('mouse events', () => {
        describe('controlledByExternalState true', () => {
            it('should not fire handleMouseEnter() & handleMouseLeave()', () => {
                const popover = mount(<Popover open={!null}>Ala</Popover>);
                const section = popover.find('section');
                const spyMouseEnter = jest.spyOn(popover.instance(), 'handleMouseEnter');
                const spyMouseLeave = jest.spyOn(popover.instance(), 'handleMouseLeave');

                section.simulate('mouseEnter', spyMouseEnter);
                expect(spyMouseEnter).not.toHaveBeenCalled();

                section.simulate('mouseLeave', spyMouseLeave);
                expect(spyMouseLeave).not.toHaveBeenCalled();
            });
        });

        describe('controlledByExternalState false', () => {
            let popover;
            let section;
            beforeEach(() => {
                popover = mount(<Popover open={null} />);
                section = popover.find('section');
            });

            it('should fire handleMouseEnter(), handleMouseLeave() & handleTransitionEnd()', () => {
                const spyMouseEnter = jest.spyOn(popover.instance(), 'handleMouseEnter');
                const spyMouseLeave = jest.spyOn(popover.instance(), 'handleMouseLeave');
                const spyTransitionEnd = jest.spyOn(popover.instance(), 'handleTransitionEnd')
                popover.instance().forceUpdate();

                section.simulate('mouseEnter', spyMouseEnter);
                expect(spyMouseEnter).toHaveBeenCalled();

                section.simulate('mouseLeave', spyMouseLeave);
                expect(spyMouseLeave).toHaveBeenCalled();

                section.simulate('transitionEnd');
                expect(spyTransitionEnd).toHaveBeenCalled();
            });

            it('should set open and renderContent to true on mouseEnter', () => {
                section.simulate('mouseEnter');
                expect(popover.state('open')).toBe(true);
                expect(popover.state('renderContent')).toBe(true);
            });

            it('should set open and renderContent to false on mouseLeave', () => {
                section.simulate('mouseLeave');
                expect(popover.state('open')).toBe(false);
                expect(popover.state('renderContent')).toBe(false);
            });

            it('should set open and renderContent to false on mouseLeave', () => {
                section.simulate('mouseLeave');
                expect(popover.state('open')).toBe(false);
                expect(popover.state('renderContent')).toBe(false);
            });
        });
    });
});

