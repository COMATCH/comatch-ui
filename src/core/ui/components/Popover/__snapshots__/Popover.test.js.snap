// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render correctly 1`] = `
<section
  className="Popover"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTransitionEnd={[Function]}
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-question-circle fa-w-16 Popover__toggle Popover__icon"
    data-icon="question-circle"
    data-prefix="fas"
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <div
    aria-hidden={true}
    className="Popover__content"
    role="tooltip"
  />
</section>
`;

exports[`Popover should render the correct classes should have a position class 1`] = `
<section
  className="Popover left"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTransitionEnd={[Function]}
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-question-circle fa-w-16 Popover__toggle Popover__icon"
    data-icon="question-circle"
    data-prefix="fas"
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <div
    aria-hidden={true}
    className="Popover__content"
    role="tooltip"
  />
</section>
`;

exports[`Popover should render the correct classes should have an open class when open is true 1`] = `
<section
  className="Popover open"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTransitionEnd={[Function]}
>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-question-circle fa-w-16 Popover__toggle Popover__icon"
    data-icon="question-circle"
    data-prefix="fas"
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M504 256c0 136.997-111.043 248-248 248S8 392.997 8 256C8 119.083 119.043 8 256 8s248 111.083 248 248zM262.655 90c-54.497 0-89.255 22.957-116.549 63.758-3.536 5.286-2.353 12.415 2.715 16.258l34.699 26.31c5.205 3.947 12.621 3.008 16.665-2.122 17.864-22.658 30.113-35.797 57.303-35.797 20.429 0 45.698 13.148 45.698 32.958 0 14.976-12.363 22.667-32.534 33.976C247.128 238.528 216 254.941 216 296v4c0 6.627 5.373 12 12 12h56c6.627 0 12-5.373 12-12v-1.333c0-28.462 83.186-29.647 83.186-106.667 0-58.002-60.165-102-116.531-102zM256 338c-25.365 0-46 20.635-46 46 0 25.364 20.635 46 46 46s46-20.636 46-46c0-25.365-20.635-46-46-46z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <div
    aria-hidden={false}
    className="Popover__content"
    role="tooltip"
  />
</section>
`;
